Unescape the URLs in the router

michalczaplinski opened the pull request:
<!--
Thanks for your pull request 😊. Note that not following the template might result in your issue being closed
-->

#### Description of what you did:

Fixes: #294 
<!-- Please describe everything as much as possible -->

#### My PR is a:

<!-- Delete the ones that don't apply -->

- 🐞 Bug fix
#### Is the PR ready to be merged?

<!-- Delete the one that don't apply -->

- ✅ Yes!

michalczaplinski committed:
tiny-router: Unescape the URLs in the router

michalczaplinski committed:
Add changeset

michalczaplinski commented:
blocked by #302 because it needs the `decode` function from `"frontity"`

luisherranz commented:
I've merged https://github.com/frontity/frontity/pull/302. Feel free to merge `dev` in this branch.

Moving this back to to-do 👍

Commit:
Merge branch 'dev' into fix-ga-links

Commit:
Merge branch 'dev' into fix-ga-links

luisherranz reviewed (commented):
I'm not sure if `decode` should be added in the `normalize` (`routeToParams`) instead of here:
https://github.com/frontity/frontity/blob/fix-ga-links/packages/wp-source/src/libraries/route-utils.ts#L40-L54

@DAreRodz what do you think?

By the way, why the links can contain an encoded query in the first place? Where does this come from?

luisherranz reviewed (changes requested):
Could you please add a couple of unit tests to cover these cases?

michalczaplinski commented:
I've just talked with Reyes, this does not actually fix the problem. 

The problematic content shows up inside of the `state.source.post[id].content.rendered`.

<img width="624" alt="Screenshot 2020-02-10 at 12 23 39" src="https://user-images.githubusercontent.com/5417266/74173491-2d2f3180-4c00-11ea-9316-77c7842f63db.png">

I think that we would need a "anchor link processor" in order to solve this...? Could you confirm that this is the right approach @luisherranz ?

luisherranz commented:
I see. Let's discuss this via Skype 👍

luisherranz commented:
Our conclusions from the call:

- WordPress does encode the links in the REST API output so we have to solve this.
- Apparently, we are already decoding the content in the html2react parser:
https://github.com/frontity/frontity/blob/fix-ga-links/packages/html2react/src/libraries/parse/index.ts#L53

So next thing to investigate why that `decode` call is not fixing the encoded links.

michalczaplinski commented:
So, I can confirm that we have not been unescaping the URLs in the anchor links. 

Also, I noticed that our `html2react` is a bit rudimental. There are some edge cases that we're not handling, like `for` attribute on `<label>` should be translated to `htmlFor` in JSX. 

React-DOM does much more when mapping between the DOM and react: https://github.com/facebook/react/blob/master/packages/react-dom/src/shared/DOMProperty.js

I will fix whatever inconsistencies I see there as well.

@luisherranz

michalczaplinski committed:
Revert "tiny-router: Unescape the URLs in the router"

This reverts commit 269fee3bc2b9c1c7276c72679639dd7cbee4c818.

michalczaplinski committed:
Revert "Add changeset"

This reverts commit 42cf7c04939b95ad9ca4e57160400acf676344ec.

michalczaplinski committed:
html2react: unescape anchor links

michalczaplinski committed:
Merge branch 'fix-ga-links' of github.com:frontity/frontity into fix-ga-links

Commit:
Merge branch 'dev' into fix-ga-links

michalczaplinski commented:
Also, this will need a couple of new tests, of course.

luisherranz commented:
> So, I can confirm that we have not been unescaping the URLs in the anchor links.

Oh, that's true, we only `decode` the content, not the attributes (props). 

What is your proposed solution then? `decode` the attributes?

michalczaplinski commented:
yup, exactly. Right now, I'm only decoding the `href`, but I'm not sure if this should apply to all other attributes by default. Do you think this is sufficient for now @luisherranz ?

luisherranz commented on packages/html2react/src/libraries/parse/index.ts:
```diff
@@ -27,6 +27,10 @@ const adaptNode: AdaptNode = (himalayaNode, parent) => {
             props.css = css(value);
+          } else if (key === "href") {
+            props.href = decode(value);
+          } else if (key === "for") {
```
What is this for?

michalczaplinski replied:
sorry, the following line should be : `props.htmlFor = value` because of https://reactjs.org/docs/dom-elements.html#htmlfor. I will fix that.

luisherranz reviewed (changes requested):
> but I'm not sure if this should apply to all other attributes by default

I don't think so because with the current method we are creating a new HTML node for each `decode` call and that is slow.

Maybe in the future we can add an option for a faster "non-full version" that implements something like https://github.com/humanmade/simple-entity-decode for these cases:
```js
decode("string", { full: false });
```

By the way, this is marked as ready to merge but two things are still missing:
- Unit tests
- Changeset

michalczaplinski commented:
> By the way, this is marked as ready to merge but two things are still missing:
> - Unit tests
> - Changeset

yup, I just wanted to confirm the solution first.

michalczaplinski committed:
html2react: Add test cases and fix types

michalczaplinski committed:
Add changeset

michalczaplinski committed:
Merge branch 'fix-ga-links' of github.com:frontity/frontity into fix-ga-links

Commit:
Merge branch 'dev' into fix-ga-links

DAreRodz commented on packages/html2react/src/libraries/parse/index.ts:
```diff
@@ -22,11 +22,21 @@ const adaptNode: AdaptNode = (himalayaNode, parent) => {
         (props: Element["props"], { key, value }) => {
+          // mapping from HTML attribute names to react names:  // https://github.com/facebook/react/blob/58b8797b7372c9296e65e08ce8297e4a394b7972/packages/react-dom/src/shared/DOMProperty.js#L241-L244
           if (key === "class") {
             props.className = value;
```
As these are special cases and are handled directly, could you remove `className`, `htmlFor`, `acceptCharset` and `httpEquiv` from [parse/attributes/html.json](https://github.com/frontity/frontity/blob/4a664272dbcccfd6b9a0eea54f40104e40c75eb7/packages/html2react/src/libraries/parse/attributes/html.json)?

DAreRodz commented on packages/html2react/src/libraries/parse/index.ts:
```diff
@@ -22,11 +22,21 @@ const adaptNode: AdaptNode = (himalayaNode, parent) => {
+          } else if (key === "style") {
             props.css = css(value);
+            // Wordpress returns links with HTML escaped entitites so we have to decode them
+          } else if (key === "href") {
```
I guess we need the same for `src` or `srcset` for images, iframes, etc, right? 🤔

michalczaplinski replied:
good point! Will add it 👍

DAreRodz reviewed (dismissed):
Code is great! 🙌

I just saw a couple of things that I think can be fixed before merging.

luisherranz commented:
I've been investigating this because I'm a bit worried about the performance of our `decode`.

1. We can do a simple regexp to detect and fix the most common entities. Something similar to what Ryan is doing here: https://github.com/humanmade/simple-entity-decode/blob/master/index.js (or even use his library).
2. We can check again if there is any entity left, using this regexp: `/&[a-zA-Z]+(;|\s)/`
    - This is the complete list of HTML entities: https://html.spec.whatwg.org/entities.json
    - You can see the regexp in action here: https://regex101.com/r/ufVAxt/2
3. If we detect a missing entity that was not fixed with the fast implementation, we apply the slower `he` (server) or the `DOMParser` (client).

Once we do that we can add `decode` to all attributes by default and avoid maintaining a list of attributes that need decoding.

@DAreRodz @michalczaplinski thoughts?

michalczaplinski commented:
@luisherranz I def like the idea! But I would actually benchmark the performance first. My expectation was that the `DOMParser` should be plenty fast, but this list of all HTML entities looks pretty long so might be wrong 😅

I ll make a quick benchmark.

luisherranz commented:
I did a benchmark and it is significantly faster: https://twitter.com/luisherranz/status/1165196400507981825 :)

michalczaplinski commented:
oh, can you share the code someplace? I'd like to compare it with using `DOMParser` as we do now on the client. 

Also, by default, does `decode` run both on the client and the server when "normally" rendering the pages?

michalczaplinski commented:
yeah, according to my quick and dirty test, `DOMParser` is super slow. At least an order of magnitude slower than `he`. The regex `/&[a-zA-Z]+(;|\s)/` is super fast, about 100x faster than `he` and 1000x faster than `DOMParser`.

<img width="625" alt="Screenshot 2020-02-14 at 15 50 22" src="https://user-images.githubusercontent.com/5417266/74566554-b9fc2700-4f41-11ea-8acb-5707b2375e99.png">

Time in miliseconds. 

This is more or less a representative run of: 

```js
i = 0;
time = performance.now();
while (i < 100) {
  // clientDecode(html) / he.decode(html) / regex(html)
  i++;
}
console.log("decode/he/regex", performance.now() - time);
```

michalczaplinski committed:
frontity: optimize `decode`

- Opt out of decoding HTML if it doesn't contain special characters
- Use a basic version of `decode` by default that ony handles most common characters.
- As a last resort, use the full decode

michalczaplinski committed:
Merge branch 'fix-ga-links' of github.com:frontity/frontity into fix-ga-links

michalczaplinski commented:
@luisherranz The method from https://github.com/humanmade/simple-entity-decode/blob/master/index.js is about 2x faster than `he`:

<img width="625" alt="Screenshot 2020-02-14 at 16 18 51" src="https://user-images.githubusercontent.com/5417266/74568317-b2d71800-4f45-11ea-995d-27846322880c.png">

I've added the optimization like you suggested with some tests.

- Opt out of decoding HTML if it doesn't contain special characters
- Use a basic version of `decode` by default that ony handles most common characters.
- As a last resort, use the full decode

luisherranz commented:
The test is just an html2react execution with x6 slowdown in the Chrome devtools.

I think the bottleneck of createElement/DomParser is the paint that the browser needs to execute (blue in the picture) but you can test it out if you want 👍

Commit:
Merge branch 'dev' into fix-ga-links

luisherranz commented:
Awesome!! 😁

luisherranz reviewed (changes requested):
The new optimization looks great! 😄 

These two things are missing:
- Remove `className`, `htmlFor`, `acceptCharset` and `httpEquiv` from parse/attributes/html.json.
- Either create a whitelist of attributes that need decoding (like `href`, `src`, `srcset` and so on...) or decode all the attributes by default.

michalczaplinski committed:
html2react: Decode all HTML attributes that are strings

michalczaplinski committed:
Merge branch 'fix-ga-links' of github.com:frontity/frontity into fix-ga-links

michalczaplinski committed:
Add changeset

michalczaplinski commented:
> Either create a whitelist of attributes that need decoding (like href, src, srcset and so on...) or decode all the attributes by default.

So, I decided to just `decode` all the attributes that are strings. I could not find a good reason not to.

Commit:
Merge branch 'dev' into fix-ga-links

luisherranz committed:
Improve changeset messages

luisherranz reviewed (approved):
I have improved the changesets because the `frontity` one was referring to the `html2react` change instead of to the performance improvement of `decode`. But other than that evertyhing is fine. Thanks @michalczaplinski!

luisherranz committed:
Merge branch 'dev' into fix-ga-links
